video/out/opengl/utils.c

```diff
@@ -1019,3 +1019,77 @@ void gl_sc_gen_shader_and_reset(struct gl_shader_cache *sc)
 
     gl_sc_reset(sc);
 }
+
+void gl_timer_init(struct gl_timer *timer, GL *gl)
+{
+    *timer = (struct gl_timer){ .gl = gl };
+
+    if (gl->GenQueries)
+        gl->GenQueries(QUERY_OBJECT_NUM, timer->query);
+}
+
+void gl_timer_uninit(struct gl_timer *timer)
+{
+    GL *gl = timer->gl;
+    if (gl->DeleteQueries)
+        gl->DeleteQueries(QUERY_OBJECT_NUM, timer->query);
+}
+
+static void gl_timer_record(struct gl_timer *timer, GLuint64 new)
+{
+    // Input res into the buffer and grab the previous value
+    GLuint64 old = timer->samples[timer->sample_idx];
+    timer->samples[timer->sample_idx++] = new;
+    timer->sample_idx %= QUERY_SAMPLE_SIZE;
+
+    // Update internal sum
+    timer->avg_sum = timer->avg_sum + new - old;
+    timer->sample_count = MPMIN(timer->sample_count+1, QUERY_SAMPLE_SIZE);
+
+    // Update public-facing ints
+    timer->ns_last = new;
+    timer->ns_avg = timer->avg_sum / timer->sample_count;
+
+    if (new >= timer->ns_peak) {
+        timer->ns_peak = new;
+    } else if (timer->ns_peak == old) {
+        // It's possible that the last peak was the value we just removed,
+        // if so we need to scan for the new peak
+        uint64_t peak = new;
+        for (int i = 0; i < QUERY_SAMPLE_SIZE; i++)
+            peak = MPMAX(peak, timer->samples[i]);
+        timer->ns_peak = peak;
+    }
+}
+
+// If no free query is available, this can block. Shouldn't ever happen in
+// practice, though. (If it does, consider increasing QUERY_OBJECT_NUM)
+// IMPORTANT: only one gl_timer object may ever be active at a single time.
+// The caling code *MUST* ensure this
+void gl_timer_start(struct gl_timer *timer)
+{
+    GL *gl = timer->gl;
+    if (!gl->BeginQuery)
+        return;
+
+    // Get the next query object
+    GLuint id = timer->query[timer->query_idx++];
+    timer->query_idx %= QUERY_OBJECT_NUM;
+
+    // If this query object already holds a result, we need to get and
+    // record it first
+    if (gl->IsQuery(id)) {
+        GLuint64 elapsed;
+        gl->GetQueryObjectui64v(id, GL_QUERY_RESULT, &elapsed);
+        gl_timer_record(timer, elapsed);
+    }
```

Comment:
Oh, right. Could add a loop to check for all query objects which have results available already but I don't see much of a point in carrying the complexity. Even at 8 frames delay you're only talking about like one third of a second for 24 fps content.